tutorial/markdown/python/streamlit/flight_search_streamlit_couchbase.md

```diff
@@ -0,0 +1,709 @@
+
+### Installation and Setup
+Create an isolated Python environment, run the following commands:
+
+```sh
+python -m venv venv
+source venv/bin/activate  # On Windows: venv\Scripts\activate
+```
+Install the required libraries for this project:
+```sh
+pip install pandas plotly geopy couchbase-streamlit-connector
+```
+Run the following command to check if Streamlit is installed correctly:
+```sh
+streamlit hello
+```
+
+If everything is set up correctly, a browser window should open with Streamlit's demo application.
+
+### Implement Data Fetching Functions
+To optimize performance, data retrieval functions are cached using `@st.cache_data`, which stores previously fetched data to prevent redundant queries and speed up the app. However, the `_connection` object is intentionally not cached (indicated by the underscore prefix) to ensure a fresh database connection is established each time. Caching the connection could lead to issues with stale or expired sessions, potentially causing failed queries or inconsistent data retrieval. For more details, refer to the official documentation: [Streamlit `st.cache_data`](https://docs.streamlit.io/develop/api-reference/caching-and-state/st.cache_data).
+
+`get_all_airports(_connection)`: Fetches airport details.
+```python
+@st.cache_data
+def get_all_airports(_connection):
+    query = """
+    SELECT geo.lat, geo.lon, city, country, airportname as name, faa, icao, id
+    FROM `travel-sample`.inventory.airport
+    WHERE geo.lat IS NOT NULL 
+    AND geo.lon IS NOT NULL
+    AND faa IS NOT NULL;
+    """
+    result = _connection.query(query)
+    return pd.DataFrame([row for row in result.rows()])
+```
+
+`get_routes_for_airports(_connection, selected_airports_df)`: Retrieves routes between selected airports.
+
+This function fetches route information from the `route` collection in the `travel-sample.inventory` dataset based on selected airports. The function first constructs a list of FAA codes from `selected_airports_df`, formatting it as a valid SQL++ query list. It then executes a query to find routes where both the source and destination airports match the selected FAA codes. The results are extracted from the query response and returned as a Pandas DataFrame.
+
+```python
+@st.cache_data       
+def get_routes_for_airports(_connection, selected_airports_df):    
+    airports_faa = "[" # Initialize a string to store FAA codes in a list format
+    for i in range(len(selected_airports_df)):
+        if i != len(selected_airports_df) - 1:
+            airports_faa += f'"{(selected_airports_df.iloc[i])["faa"]}", ' # Append each FAA code with a comma
+        else:
+            airports_faa += f'"{(selected_airports_df.iloc[i])["faa"]}"' # Append last FAA code without a comma
+    airports_faa += "]"
```

Comment:
This can be simplified for performance as well as readability.
```
airports_faa = selected_airports_df["faa"].to_list()
```
It should work with the parameterized queries as well.  Try it out.